File: GOF/Prototype/java/src/MazePrototypeFactory.java
import maze.Door;
import maze.Maze;
import maze.Room;
import maze.Wall;

public class MazePrototypeFactory extends MazeFactory {

	private Maze m_mazePrototype;
	private Wall m_wallPrototype;
	private Room m_roomPrototype;
	private Door m_doorPrototype;

	MazePrototypeFactory(Maze maze, Wall wall, Room room, Door door) {
		m_mazePrototype = maze;
		m_wallPrototype = wall;
		m_roomPrototype = room;
		m_doorPrototype = door;
	}

	@Override
	public Maze makeMaze() {
		try {
			return (Maze)m_mazePrototype.clone();
		} catch (CloneNotSupportedException e) {
			return null;
		}
	}

	@Override
	public Wall makeWall() {
		try {
			return (Wall) m_wallPrototype.clone();
		} catch (CloneNotSupportedException e) {
			return null;
		}
	}

	@Override
	public Room makeRoom(int n) {
		try {
			Room room = (Room)m_roomPrototype.clone();
			room.initialize(n);
			return room;
		} catch (CloneNotSupportedException e) {
			return null;
		}

	}

	@Override
	public Door makeDoor(Room r1, Room r2) {
		try {
			Door door = (Door) m_doorPrototype.clone();
			door.initialize(r1, r2);
			return door;
		} catch (CloneNotSupportedException e) {
			return null;
		}
	}
}

File: GOF/Prototype/cpp/MazePrototypeFactory.cpp
//
// Created by 17271 on 2019/3/14.
//

#include "MazePrototypeFactory.h"
#include "maze/maze.h"
#include "maze/wall.h"
#include "maze/room.h"
#include "maze/door.h"

using std::shared_ptr;              using std::make_shared;

MazePrototypeFactory::MazePrototypeFactory(std::shared_ptr<Maze> maze, std::shared_ptr<Wall> wall,
                                           std::shared_ptr<Room> room, std::shared_ptr<Door> door)
    : m_prototypeMaze{maze}, m_prototypeWall{wall}, m_prototypeRoom{room}, m_prototypeDoor{door}
{}

std::shared_ptr<Maze> MazePrototypeFactory::makeMaze() const
{
    return m_prototypeMaze->clone();
}

std::shared_ptr<Wall> MazePrototypeFactory::makeWall() const
{
    return std::dynamic_pointer_cast<Wall>(m_prototypeWall->clone());
}

std::shared_ptr<Room> MazePrototypeFactory::makeRoom(int n) const
{
    shared_ptr<Room> room = std::dynamic_pointer_cast<Room>(m_prototypeRoom->clone());
    room->initialize(n);
    return room;
}

std::shared_ptr<Door> MazePrototypeFactory::makeDoor(std::weak_ptr<Room> r1, std::weak_ptr<Room> r2)
{
    shared_ptr<Door> door = std::dynamic_pointer_cast<Door>(m_prototypeDoor->clone());
    door->initialize(r1, r2);
    return door;
}